### ISSUE TITLE
Refresh token apparently not being revoked after being used to create a new access token

### ISSUE THREAD

**Issue by thenickcox** [doorkeeper-gem/doorkeeper#1396]
### Steps to reproduce
TL;DR, I've written a spec that asserts that using after using a refresh token to generate an access token, it is possible to re-use that refresh token to generate another access token, suggesting that refresh tokens aren't revoked after use. I asked a question about this [on StackOverflow](https://stackoverflow.com/questions/58271861/doorkeeper-refresh-tokens-not-expiring-on-reuse) and didn't get a single answer.

Am I doing something wrong?

EDIT:
I'd already looked through the issues, but recently found https://github.com/doorkeeper-gem/doorkeeper/issues/1058, which _seems_ to say something similar (hard to tell, TBQH), but it was closed without any further action.

## Repro steps
1. Generate an access token (a1) and refresh token (r1)
2. Use the refresh token (r1) to generate a second access token (a2) and refresh token (r2)
3. Use the new refresh token (r2) to generate a third access token (a3) and refresh token (r3)
4. Use the first refresh token (r1) to generate another access token

Observed:
Request succeeds and an access token is generated

Expected:
Request fails because refresh token has been revoked

Failing spec:

```ruby
describe 'OAuth flow' do
  # ...
  describe 'refresh tokens' do

    # ...
    context 'when attempting reuse of a refresh token' do
      before do
        redirect_uri = 'https://localhost:3002'

        # ivars necessary brecause plaintext tokens/secrets are only available upon creation of the object
        @client = Doorkeeper::Application.create(name: 'foo', uid: 'bar', redirect_uri: redirect_uri, scopes: 'project_index')
        @secret = @client.plaintext_secret
        @grant = Doorkeeper::AccessGrant.create(resource_owner_id: user.id, organization_id: org.id, application_id: @client.id, scopes: 'project_index', expires_in: 600, redirect_uri: redirect_uri)
        @code = @grant.plaintext_token
      end

      it 'revokes the initial refresh token' do
        # Get an initial access token and refresh token
        post "/oauth/token?client_id=#{@client.uid}&client_secret=#{@secret}&code=#{@code}&grant_type=authorization_code&redirect_uri=#{@client.redirect_uri}&scope=project_index"
        expect(response.status).to eq(200)

        # Use refresh token to get a new access token
        previous_refresh_token = JSON.parse(response.body)['refresh_token']

        post "/oauth/token?client_id=#{@client.uid}&client_secret=#{@secret}&code=#{@code}&grant_type=refresh_token&refresh_token=#{previous_refresh_token}&redirect_uri=#{@client.redirect_uri}&scope=project_index"
        expect(response.status).to eq(200)
        json_body = JSON.parse(response.body)
        new_access_token = json_body['access_token']
        new_refresh_token = json_body['refresh_token']

        # Use the new access token successfully
        get "/api/v1/projects?access_token=#{new_access_token}"
        expect(response.status).to eq(200)

        # Use the new refresh token to get yet another access token
        post "/oauth/token?client_id=#{@client.uid}&client_secret=#{@secret}&code=#{@code}&grant_type=refresh_token&refresh_token=#{new_refresh_token}&redirect_uri=#{@client.redirect_uri}&scope=project_index"
        expect(response.status).to eq(200)

        # Attempt reuse of the first refresh token
        post "/oauth/token?client_id=#{@client.uid}&client_secret=#{@secret}&code=#{@code}&grant_type=refresh_token&refresh_token=#{previous_refresh_token}&redirect_uri=#{@client.redirect_uri}&scope=project_index"
        expect(response.status).to eq(400)

        # ^^^ fails, response is 200 and a new access token and refresh token are generated
      end
    end
  end

  # ...
end
```

### Expected behavior
Refresh tokens are able to be used to generate one and only one access token.

### Actual behavior
If this spec is correct, it appears that refresh tokens may be used multiple times.

### System configuration
Doorkeeper 5.2.1

**Doorkeeper initializer**:

```ruby
# config/initializers/doorkeeper.rb
# frozen_string_literal: true


doorkeeper_extend_dir = File.join(Rails.root, "app", "models", "doorkeeper")
Dir.glob("#{doorkeeper_extend_dir}/*.rb").each { |f| require(f) }

Doorkeeper.configure do
  # Change the ORM that doorkeeper will use (needs plugins)
  orm :active_record

  # This block will be called to check whether the resource owner is authenticated or not.
  resource_owner_authenticator do
    auth = Authentication.new
    token = auth.get_auth_token_from_request(request)
    session = auth.session_from_token(token)
    if session
      %i(redirect_uri oauth_client_id oauth_referred oauth_scope state).map {|k| request.session.delete(k) }
      User.find(session.instance_variable_get(:@user_id))
    else
      request.session[:redirect_uri] = params[:redirect_uri]
      request.session[:oauth_client_id] = params[:client_id]
      request.session[:oauth_scope] = params[:scope]
      request.session[:oauth_referred] = true
      request.session[:state] = params[:state] if params[:state]
      redirect_uri = "#{request.base_url}#{sign_in_path}?redirect_uri=#{params[:redirect_uri]}"
      redirect_uri += "&state=#{params[:state]}" if params[:state]
      redirect_to(redirect_uri)
    end
  end

  # If you didn't skip applications controller from Doorkeeper routes in your application routes.rb
  # file then you need to declare this block in order to restrict access to the web interface for
  # adding oauth authorized applications. In other case it will return 403 Forbidden response
  # every time somebody will try to access the admin web interface.
  #
  admin_authenticator do
    # Admin auth logic
    auth = Authentication.new
    authToken = auth.get_auth_token_from_request(request)
    session = auth.session_from_token(authToken)

    if session
      head :forbidden unless session && session.is_admin?
    else
      redirect_to si_path
    end
  end
  
  hash_token_secrets
  
  hash_application_secrets using: '::Doorkeeper::SecretStoring::BCrypt'

  use_refresh_token
end
```

**Ruby version**: ``
ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-linux]

**Gemfile.lock**:
  
<details>
```
GIT
  remote: git://github.com/10Kft/elastic_whenever.git
  revision: 9d5dc54166d8a78dd83a5d6aaf488e4823876c48
  branch: fix-rule-name
  specs:
    elastic_whenever (0.3.5)
      aws-sdk-cloudwatchevents (~> 1.0)
      aws-sdk-ecs (~> 1.0)
      aws-sdk-iam (~> 1.0)
      chronic (~> 0.10)

GIT
  remote: git://github.com/10Kft/octopus.git
  revision: 2e5c0eeac41fe2f1ced498dc751a6c35e347576f
  branch: fix/current-shard
  specs:
    ar-octopus (0.9.1)
      activerecord (>= 4.0.0)
      activesupport (>= 4.0.0)

GIT
  remote: git://github.com/tcocca/acts_as_follower.git
  revision: c5ac7b9601c4af01eb4d9112330b27be4d694ecc
  branch: master
  specs:
    acts_as_follower (0.2.1)
      activerecord (>= 4.0)

GIT
  remote: https://github.com/10Kft/ruby-saml
  revision: 6ba57b93889b739299d1970e7b155992c6c31500
  branch: tenkft-frozen
  specs:
    ruby-saml (0.8.1)
      nokogiri (>= 1.5.0)
      uuid (~> 2.3)

GEM
  remote: https://rubygems.org/
  remote: https://gems.contribsys.com/
  specs:
    CFPropertyList (2.3.6)
    actioncable (5.1.6)
      actionpack (= 5.1.6)
      nio4r (~> 2.0)
      websocket-driver (~> 0.6.1)
    actionmailer (5.1.6)
      actionpack (= 5.1.6)
      actionview (= 5.1.6)
      activejob (= 5.1.6)
      mail (~> 2.5, >= 2.5.4)
      rails-dom-testing (~> 2.0)
    actionpack (5.1.6)
      actionview (= 5.1.6)
      activesupport (= 5.1.6)
      rack (~> 2.0)
      rack-test (>= 0.6.3)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    actionview (5.1.6)
      activesupport (= 5.1.6)
      builder (~> 3.1)
      erubi (~> 1.4)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.3)
    active_record_doctor (1.5.0)
      activerecord (>= 4.2)
      railties (>= 4.2)
    activejob (5.1.6)
      activesupport (= 5.1.6)
      globalid (>= 0.3.6)
    activemodel (5.1.6)
      activesupport (= 5.1.6)
    activerecord (5.1.6)
      activemodel (= 5.1.6)
      activesupport (= 5.1.6)
      arel (~> 8.0)
    activesupport (5.1.6)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (>= 0.7, < 2)
      minitest (~> 5.1)
      tzinfo (~> 1.1)
    addressable (2.5.2)
      public_suffix (>= 2.0.2, < 4.0)
    annotate (2.7.2)
      activerecord (>= 3.2, < 6.0)
      rake (>= 10.4, < 13.0)
    arel (8.0.0)
    ast (2.4.0)
    attr_encrypted (3.1.0)
      encryptor (~> 3.0.0)
    awrence (1.1.0)
    aws-eventstream (1.0.2)
    aws-partitions (1.149.0)
    aws-sdk (1.56.0)
      aws-sdk-v1 (= 1.56.0)
    aws-sdk-cloudwatchevents (1.17.0)
      aws-sdk-core (~> 3, >= 3.48.2)
      aws-sigv4 (~> 1.1)
    aws-sdk-core (3.48.3)
      aws-eventstream (~> 1.0, >= 1.0.2)
      aws-partitions (~> 1.0)
      aws-sigv4 (~> 1.1)
      jmespath (~> 1.0)
    aws-sdk-ecs (1.36.0)
      aws-sdk-core (~> 3, >= 3.48.2)
      aws-sigv4 (~> 1.1)
    aws-sdk-iam (1.18.0)
      aws-sdk-core (~> 3, >= 3.48.2)
      aws-sigv4 (~> 1.1)
    aws-sdk-v1 (1.56.0)
      json (~> 1.4)
      nokogiri (>= 1.4.4)
    aws-sigv4 (1.1.0)
      aws-eventstream (~> 1.0, >= 1.0.2)
    bcrypt (3.1.13)
    better_errors (2.4.0)
      coderay (>= 1.0.0)
      erubi (>= 1.0.0)
      rack (>= 0.9.0)
    binding_of_caller (0.8.0)
      debug_inspector (>= 0.0.1)
    bootsnap (1.2.0)
      msgpack (~> 1.0)
    builder (3.2.3)
    bulk_insert (1.6.0)
      activerecord (>= 4.1.0)
    bullet (5.7.5)
      activesupport (>= 3.0.0)
      uniform_notifier (~> 1.11.0)
    byebug (10.0.0)
    childprocess (0.9.0)
      ffi (~> 1.0, >= 1.0.11)
    choice (0.2.0)
    chronic (0.10.2)
    coderay (1.1.2)
    concurrent-ruby (1.0.5)
    connection_pool (2.2.2)
    crass (1.0.4)
    css_parser (1.6.0)
      addressable
    daemons (1.3.1)
    dante (0.2.0)
    database_cleaner (1.6.2)
    ddtrace (0.24.0)
      msgpack
    debug_inspector (0.0.3)
    diff-lcs (1.3)
    docile (1.1.5)
    dogapi (1.28.0)
      multi_json
    dogstatsd-ruby (3.3.0)
    doorkeeper (5.2.1)
      railties (>= 5)
    encryptor (3.0.0)
    erubi (1.7.1)
    eventmachine (1.2.5)
    excon (0.60.0)
    factory_bot (4.8.2)
      activesupport (>= 3.0.0)
    factory_bot_rails (4.8.2)
      factory_bot (~> 4.8.2)
      railties (>= 3.0.0)
    faker (1.8.7)
      i18n (>= 0.7)
    faraday (0.13.1)
      multipart-post (>= 1.2, < 3)
    faraday_middleware (0.10.1)
      faraday (>= 0.7.4, < 1.0)
    faye-websocket (0.10.7)
      eventmachine (>= 0.12.0)
      websocket-driver (>= 0.5.1)
    ffi (1.9.23)
    fog-aws (2.0.1)
      fog-core (~> 1.38)
      fog-json (~> 1.0)
      fog-xml (~> 0.1)
      ipaddress (~> 0.8)
    fog-core (1.45.0)
      builder
      excon (~> 0.58)
      formatador (~> 0.2)
    fog-json (1.0.2)
      fog-core (~> 1.0)
      multi_json (~> 1.10)
    fog-xml (0.1.3)
      fog-core
      nokogiri (>= 1.5.11, < 2.0.0)
    foreman (0.84.0)
      thor (~> 0.19.1)
    formatador (0.2.5)
    gem_bench (1.0.4)
      bundler (~> 1.14)
    get_process_mem (0.2.1)
    gli (2.17.1)
    globalid (0.4.1)
      activesupport (>= 4.2.0)
    guard (2.14.2)
      formatador (>= 0.2.4)
      listen (>= 2.7, < 4.0)
      lumberjack (>= 1.0.12, < 2.0)
      nenv (~> 0.1)
      notiffany (~> 0.0)
      pry (>= 0.9.12)
      shellany (~> 0.0)
      thor (>= 0.18.1)
    guard-compat (1.2.1)
    guard-rspec (4.7.3)
      guard (~> 2.1)
      guard-compat (~> 1.1)
      rspec (>= 2.99.0, < 4.0)
    hashie (3.5.7)
    http-exceptions (0.1.2)
    httparty (0.16.1)
      multi_xml (>= 0.5.2)
    i18n (1.1.0)
      concurrent-ruby (~> 1.0)
    icalendar (2.4.1)
    ipaddress (0.8.3)
    jmespath (1.4.0)
    json (1.8.6)
    kramdown (1.16.2)
    libv8 (6.3.292.48.1)
    libv8 (6.3.292.48.1-x86_64-darwin-17)
    listen (3.1.5)
      rb-fsevent (~> 0.9, >= 0.9.4)
      rb-inotify (~> 0.9, >= 0.9.7)
      ruby_dep (~> 1.2)
    lograge (0.9.0)
      actionpack (>= 4)
      activesupport (>= 4)
      railties (>= 4)
      request_store (~> 1.0)
    loofah (2.2.2)
      crass (~> 1.0.2)
      nokogiri (>= 1.5.9)
    lumberjack (1.0.12)
    macaddr (1.7.1)
      systemu (~> 2.6.2)
    mail (2.7.0)
      mini_mime (>= 0.1.1)
    mem_info (1.0.1)
    memory_profiler (0.9.10)
    metaclass (0.0.4)
    method_source (0.8.2)
    mime-types (3.1)
      mime-types-data (~> 3.2015)
    mime-types-data (3.2016.0521)
    mini_mime (1.0.1)
    mini_portile2 (2.3.0)
    minitest (5.11.3)
    mocha (1.3.0)
      metaclass (~> 0.0.1)
    msgpack (1.2.4)
    multi_json (1.13.1)
    multi_xml (0.6.0)
    multipart-post (2.1.1)
    mysql2 (0.4.10)
    nenv (0.3.0)
    nio4r (2.3.1)
    nokogiri (1.8.4)
      mini_portile2 (~> 2.3.0)
    notiffany (0.1.1)
      nenv (~> 0.1)
      shellany (~> 0.0)
    oj (3.5.0)
    parallel (1.12.1)
    parallel_tests (2.21.2)
      parallel
    parser (2.5.0.3)
      ast (~> 2.4.0)
    pdf-core (0.7.0)
    plissken (1.3.0)
    power_assert (1.1.1)
    powerpack (0.1.1)
    prawn (2.2.2)
      pdf-core (~> 0.7.0)
      ttfunk (~> 1.5)
    protected_attributes_continued (1.3.0)
      activemodel (~> 5.0)
    pry (0.10.4)
      coderay (~> 1.1.0)
      method_source (~> 0.8.1)
      slop (~> 3.4)
    pry-nav (0.2.4)
      pry (>= 0.9.10, < 0.11.0)
    pry-rails (0.3.6)
      pry (>= 0.10.4)
    public_suffix (3.0.2)
    puma (3.11.3)
    puma_worker_killer (0.1.0)
      get_process_mem (~> 0.2)
      puma (>= 2.7, < 4)
    rabl (0.13.1)
      activesupport (>= 2.3.14)
    rack (2.0.5)
    rack-attack (5.1.0)
      rack
    rack-cache (1.7.2)
      rack (>= 0.4)
    rack-cors (1.0.2)
    rack-mini-profiler (1.0.1)
      rack (>= 1.2.0)
    rack-protection (2.0.5)
      rack
    rack-test (1.1.0)
      rack (>= 1.0, < 3)
    rails (5.1.6)
      actioncable (= 5.1.6)
      actionmailer (= 5.1.6)
      actionpack (= 5.1.6)
      actionview (= 5.1.6)
      activejob (= 5.1.6)
      activemodel (= 5.1.6)
      activerecord (= 5.1.6)
      activesupport (= 5.1.6)
      bundler (>= 1.3.0)
      railties (= 5.1.6)
      sprockets-rails (>= 2.0.0)
    rails-controller-testing (1.0.2)
      actionpack (~> 5.x, >= 5.0.1)
      actionview (~> 5.x, >= 5.0.1)
      activesupport (~> 5.x)
    rails-dom-testing (2.0.3)
      activesupport (>= 4.2.0)
      nokogiri (>= 1.6)
    rails-erd (1.5.2)
      activerecord (>= 3.2)
      activesupport (>= 3.2)
      choice (~> 0.2.0)
      ruby-graphviz (~> 1.2)
    rails-html-sanitizer (1.0.4)
      loofah (~> 2.2, >= 2.2.2)
    rails-observers (0.1.5)
      activemodel (>= 4.0)
    railties (5.1.6)
      actionpack (= 5.1.6)
      activesupport (= 5.1.6)
      method_source
      rake (>= 0.8.7)
      thor (>= 0.18.1, < 2.0)
    rainbow (3.0.0)
    rake (12.3.1)
    rb-fsevent (0.10.3)
    rb-inotify (0.9.10)
      ffi (>= 0.5.0, < 2)
    redis (4.1.2)
    redis-actionpack (5.0.2)
      actionpack (>= 4.0, < 6)
      redis-rack (>= 1, < 3)
      redis-store (>= 1.1.0, < 2)
    redis-activesupport (5.0.4)
      activesupport (>= 3, < 6)
      redis-store (>= 1.3, < 2)
    redis-prescription (1.0.0)
    redis-rack (2.0.4)
      rack (>= 1.5, < 3)
      redis-store (>= 1.2, < 2)
    redis-rack-cache (2.0.2)
      rack-cache (>= 1.6, < 2)
      redis-store (>= 1.2, < 2)
    redis-rails (5.0.2)
      redis-actionpack (>= 5.0, < 6)
      redis-activesupport (>= 5.0, < 6)
      redis-store (>= 1.2, < 2)
    redis-store (1.5.0)
      redis (>= 2.2, < 5)
    request-log-analyzer (1.13.4)
    request_store (1.4.0)
      rack (>= 1.4)
    responders (2.4.0)
      actionpack (>= 4.2.0, < 5.3)
      railties (>= 4.2.0, < 5.3)
    roadie (3.2.2)
      css_parser (~> 1.4)
      nokogiri (~> 1.5)
    rollbar (2.18.0)
      multi_json
    rspec (3.7.0)
      rspec-core (~> 3.7.0)
      rspec-expectations (~> 3.7.0)
      rspec-mocks (~> 3.7.0)
    rspec-activemodel-mocks (1.0.3)
      activemodel (>= 3.0)
      activesupport (>= 3.0)
      rspec-mocks (>= 2.99, < 4.0)
    rspec-core (3.7.1)
      rspec-support (~> 3.7.0)
    rspec-expectations (3.7.0)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.7.0)
    rspec-instafail (1.0.0)
      rspec
    rspec-mocks (3.7.0)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.7.0)
    rspec-rails (3.7.2)
      actionpack (>= 3.0)
      activesupport (>= 3.0)
      railties (>= 3.0)
      rspec-core (~> 3.7.0)
      rspec-expectations (~> 3.7.0)
      rspec-mocks (~> 3.7.0)
      rspec-support (~> 3.7.0)
    rspec-retry (0.5.6)
      rspec-core (> 3.3, < 3.8)
    rspec-support (3.7.1)
    rubocop (0.53.0)
      parallel (~> 1.10)
      parser (>= 2.5)
      powerpack (~> 0.1)
      rainbow (>= 2.2.2, < 4.0)
      ruby-progressbar (~> 1.7)
      unicode-display_width (~> 1.0, >= 1.0.1)
    rubocop-rspec (1.24.0)
      rubocop (>= 0.53.0)
    ruby-graphviz (1.2.3)
    ruby-prof (0.17.0)
    ruby-progressbar (1.9.0)
    ruby_dep (1.5.0)
    rubyzip (1.2.1)
    scout_apm (2.4.9)
    selenium-webdriver (3.11.0)
      childprocess (~> 0.5)
      rubyzip (~> 1.2)
    shellany (0.0.1)
    sidekiq (5.2.7)
      connection_pool (~> 2.2, >= 2.2.2)
      rack (>= 1.5.0)
      rack-protection (>= 1.5.0)
      redis (>= 3.3.5, < 5)
    sidekiq-pro (4.0.5)
      concurrent-ruby (>= 1.0.5)
      sidekiq (>= 5.2.1)
    sidekiq-throttled (0.10.0)
      concurrent-ruby
      redis-prescription
      sidekiq
    simplecov (0.15.1)
      docile (~> 1.1.0)
      json (>= 1.8, < 3)
      simplecov-html (~> 0.10.0)
    simplecov-html (0.10.2)
    slack-bot-server (0.4.7)
      faye-websocket (~> 0.10)
      multi_json
      slack-ruby-client (~> 0.7)
    slack-ruby-client (0.11.1)
      activesupport
      faraday (>= 0.9)
      faraday_middleware
      gli
      hashie
      websocket-driver
    slop (3.6.0)
    smartsheet (2.77.2)
      awrence (~> 1.0)
      faraday (~> 0.13.1)
      faraday_middleware (~> 0.10.0)
      plissken (~> 1.2)
    sprockets (3.7.2)
      concurrent-ruby (~> 1.0)
      rack (> 1, < 3)
    sprockets-rails (3.2.1)
      actionpack (>= 4.0)
      activesupport (>= 4.0)
      sprockets (>= 3.0.0)
    statsd-ruby (1.4.0)
    stripe (3.11.0)
      faraday (~> 0.10)
    stripe-ruby-mock (2.5.3)
      dante (>= 0.2.0)
      multi_json (~> 1.0)
      stripe (>= 2.0.3)
    sys-proctable (1.2.0)
      ffi
    systemu (2.6.5)
    terminal-notifier-guard (1.7.0)
    test-unit (3.2.7)
      power_assert
    thor (0.19.4)
    thread_safe (0.3.6)
    timecop (0.9.1)
    transaction_isolation (1.0.3)
      activerecord (>= 3.0.11)
    ttfunk (1.5.1)
    tzinfo (1.2.5)
      thread_safe (~> 0.1)
    unf (0.1.4)
      unf_ext
    unf_ext (0.0.7.5)
    unicode-display_width (1.3.0)
    uniform_notifier (1.11.0)
    uuid (2.3.8)
      macaddr (~> 1.0)
    vcr (5.0.0)
    websocket-driver (0.6.5)
      websocket-extensions (>= 0.1.0)
    websocket-extensions (0.1.3)
    whenever (0.10.0)
      chronic (>= 0.6.3)

PLATFORMS
  ruby
  x86_64-darwin-17

DEPENDENCIES
  CFPropertyList (= 2.3.6)
  active_record_doctor
  acts_as_follower!
  annotate
  ar-octopus!
  attr_encrypted
  aws-sdk (= 1.56.0)
  bcrypt
  better_errors
  binding_of_caller
  bootsnap
  bulk_insert
  bullet
  byebug
  daemons
  database_cleaner
  ddtrace
  dogapi
  dogstatsd-ruby (>= 3.0.0)
  doorkeeper (~> 5.2.1)
  elastic_whenever!
  factory_bot_rails
  faker
  fog-aws
  foreman
  gem_bench
  guard
  guard-rspec
  http-exceptions
  httparty
  icalendar
  kramdown
  libv8
  listen
  lograge
  mem_info
  memory_profiler
  mime-types
  mocha
  mysql2 (>= 0.4.9)
  oj
  parallel_tests
  prawn
  protected_attributes_continued (>= 1.3.0)
  pry
  pry-nav
  pry-rails
  puma
  puma_worker_killer
  rabl
  rack-attack
  rack-cors
  rack-mini-profiler (~> 1.0.1)
  rails (>= 5.1.6)
  rails-controller-testing
  rails-erd
  rails-observers
  redis (= 4.1.2)
  redis-rack-cache
  redis-rails
  request-log-analyzer
  responders
  roadie
  rollbar (= 2.18.0)
  rspec-activemodel-mocks
  rspec-instafail
  rspec-rails
  rspec-retry
  rubocop
  rubocop-rspec
  ruby-prof
  ruby-saml!
  scout_apm
  selenium-webdriver
  sidekiq (= 5.2.7)
  sidekiq-pro!
  sidekiq-throttled
  simplecov
  slack-bot-server
  smartsheet (~> 2.77.0)
  statsd-ruby
  stripe
  stripe-ruby-mock (~> 2.5.0)
  sys-proctable
  terminal-notifier-guard
  test-unit
  timecop
  transaction_isolation
  unf
  vcr
  whenever

BUNDLED WITH
   1.16.1
```
</details>

**Comment by nbulaj** [comment]
Hi @thenickcox 

What we have in RFC: https://tools.ietf.org/html/rfc6749#section-6

>  The authorization server MAY issue a new refresh token, in which case
>   the client MUST discard the old refresh token and replace it with the
>   new refresh token.  The authorization server MAY revoke the old
>   refresh token after issuing a new refresh token to the client.  If a
>   new refresh token is issued, the refresh token scope MUST be
>   identical to that of the refresh token included by the client in the
>   request.

So this is what we must implement in not implemented yet. I'll revisit the code to fix the issue, thanks!

**Comment by nbulaj** [comment]
I checked our code-base and here what we have:

```
      # If there is a previous_refresh_token column,
      # refresh tokens will be revoked after a related access token is used.
      # If there is no previous_refresh_token column,
      # previous tokens are revoked as soon as a new access token is created.
```

It's from migration template. Also we have a rails generator for this column.
So if you add `previous_refresh_token` to `oauth_access_tokens` - Doorkeeper will revoke the token used to refresh every time.

It was like that before me and originally introduced here https://github.com/doorkeeper-gem/doorkeeper/pull/769

**Comment by nbulaj** [comment]
Oh, I re-read the RFC and see:

> The authorization server MAY issue a new refresh token, in which case
the **client MUST discard** the old refresh token and replace it with the
new refresh token. The authorization server MAY revoke the old
refresh token after issuing a new refresh token to the client. If a
new refresh token is issued, the refresh token scope MUST be
identical to that of the refresh token included by the client in the
request.

So auth server _"MAY issue a new refresh token"_ (that's exactly what Doorkeeper does) and **client MUST discard** the old refresh token.

Also _"authorization server MAY revoke the old refresh token after issuing a new refresh token to the client_" - this is also exactly what Doorkeeper does if you enable `previous_refresh_token` (run `  rails generate doorkeeper:previous_refresh_token`)

**Comment by thenickcox** [comment]
Hi, @nbulaj. Thanks a lot for the reply. I think I'm confused, though. I _do_ have a `previous_refresh_token` column. I also confirmed that it's getting populated.

One thing I noticed, however, is that for a particular `resource_owner_id`, the `previous_refresh_token` column value is always the same. Is that expected? I would think that if you generated an access token and refresh token (a1 & r1), and then used the refresh token (r1) to generate another access token and refresh token (a2 & r2), then used r2 to generate a third access token and refresh (a3 & r3), that the value for `previous_refresh_token` on the a3 record would be r2, but in my database, it's showing r1. Is that expected?

**Comment by nbulaj** [comment]
>  I do have a previous_refresh_token column. I also confirmed that it's getting populated.

Than it must revoke the token on next use or somehow like that, you could check the test suite to find what's going on.

> Is that expected?

I need to go through the code to check that, I don't know.

**Comment by thenickcox** [comment]
@nbulaj What I'm finding in the spec is a [`context` entitled "refresh tokens expire on access token use"](https://github.com/doorkeeper-gem/doorkeeper/blob/master/spec/lib/oauth/refresh_token_request_spec.rb#L90-L111). It's a bit confusing, though, because one of the `it`s is "does not revoke the previous token". I assume that's because the new access token hasn't been used yet. But in my spec above, I use the refresh token to generate a new access token, use that new access token, then assert that the previous refresh token has been revoked, but it hasn't. And I actually don't see a case in the gem's `refresh_token_request_spec.rb` where it uses a refresh token and then attempts to _reuse_ it.

**Comment by thenickcox** [comment]
I also see this:

```ruby
  it "revokes the previous token" do
    expect { subject.authorize }.to change { refresh_token.revoked? }.from(false).to(true)
  end
```

This is a bit misleading, though, because `refresh_token` here is an instance of `Doorkeeper::AccessToken`. It's not actually a refresh token.

**Comment by nbulaj** [comment]
hey, misunderstanding with the migration actually:

> Comment out this line if you'd rather have refresh tokens instantly revoked.

So you need to **drop the column** to get tokens to be revoked on use.

I've commented the line in spec/dummy/db/schema.rb:

```ruby
 # t.string "previous_refresh_token", default: "", null: false
```

Then added your test example from above:

```ruby
  context 'when attempting reuse of a refresh token' do
    before do
      redirect_uri = 'https://localhost:3002'
      @resource_owner = FactoryBot.create(:doorkeeper_testing_user)
      # ivars necessary brecause plaintext tokens/secrets are only available upon creation of the object
      @client = Doorkeeper::Application.create(name: 'foo', redirect_uri: redirect_uri, scopes: 'project_index')
      @secret = @client.plaintext_secret
      @grant = Doorkeeper::AccessGrant.create(resource_owner_id: @resource_owner.id, application_id: @client.id, scopes: 'project_index', expires_in: 600, redirect_uri: redirect_uri)
      @code = @grant.plaintext_token
    end

    it 'revokes the initial refresh token' do
      # Get an initial access token and refresh token
      post "/oauth/token?client_id=#{@client.uid}&client_secret=#{@secret}&code=#{@code}&grant_type=authorization_code&redirect_uri=#{@client.redirect_uri}&scope=project_index"
      expect(response.status).to eq(200)

      # Use refresh token to get a new access token
      previous_refresh_token = JSON.parse(response.body)['refresh_token']

      post "/oauth/token?client_id=#{@client.uid}&client_secret=#{@secret}&code=#{@code}&grant_type=refresh_token&refresh_token=#{previous_refresh_token}&redirect_uri=#{@client.redirect_uri}&scope=project_index"
      expect(response.status).to eq(200)
      json_body = JSON.parse(response.body)
      new_access_token = json_body['access_token']
      new_refresh_token = json_body['refresh_token']

      # Use the new access token successfully
      # get "/api/v1/projects?access_token=#{new_access_token}"
      #  expect(response.status).to eq(200)
      #
      # Use the new refresh token to get yet another access token
      post "/oauth/token?client_id=#{@client.uid}&client_secret=#{@secret}&code=#{@code}&grant_type=refresh_token&refresh_token=#{new_refresh_token}&redirect_uri=#{@client.redirect_uri}&scope=project_index"
      expect(response.status).to eq(200)

      # Attempt reuse of the first refresh token
      post "/oauth/token?client_id=#{@client.uid}&client_secret=#{@secret}&code=#{@code}&grant_type=refresh_token&refresh_token=#{previous_refresh_token}&redirect_uri=#{@client.redirect_uri}&scope=project_index"
      expect(response.status).to eq(400)
    end
  end
```

And it passes

![image](https://user-images.githubusercontent.com/1443426/79135279-c53fc780-7db7-11ea-9651-3bec31056987.png)

Sorry for confusion with the migration.

---

About 

> refresh_token here is an instance of Doorkeeper::AccessToken. It's not actually a refresh token. 

In terms of Doorkeeper Refresh Token and Access Token is the same model

**Comment by thenickcox** [comment]
Thanks so much for testing that out. But I'm confused on one other bit of that migration. The line _before_ the one you quoted is

```
Doorkeeper implementation: if there is a `previous_refresh_token` column,
refresh tokens will be revoked after a related access token is used.
```

So in my implementation, I have that `previous_refresh_token`, but my test generates an access token with the refresh token and it isn't expired. In other words, it seems like that test should pass _without_ that column.

And that'd be the preferred behavior. Because without that column, previous tokens are revoked as soon as the new access token is created. But what if the response with the new access token fails? Then the client can't try again with the refresh token, and won't have access to the access token.

**Comment by thenickcox** [comment]
@nbulaj Update here: someone in the [Stack Overflow conversation](https://stackoverflow.com/questions/58271861/doorkeeper-refresh-tokens-not-expiring-on-reuse) pointed me toward [this pull request](https://github.com/doorkeeper-gem/doorkeeper/pull/1341/files), which apparently fixed `revoked_on_use` when `hash_token_secrets` was enabled.

I do have `hash_token_secrets` enabled. So I upgraded to 5.3.1 to get that change, but I'm still seeing this test fail, and I'm still seeing the ability to re-use a refresh token that should be revoked.

**Comment by khiav223577** [comment]
It looks like this issue still exists in doorkeeper v5.8.0

**Comment by reptilianking** [comment]
> Revisé nuestra base de código y esto es lo que tenemos:
> 
> ```
>       # If there is a previous_refresh_token column,
>       # refresh tokens will be revoked after a related access token is used.
>       # If there is no previous_refresh_token column,
>       # previous tokens are revoked as soon as a new access token is created.
> ```
> 
> Proviene de la plantilla de migración. También contamos con un generador de Rails para esta columna. Por lo tanto, si agregas " `previous_refresh_token`Totalizador `oauth_access_tokens`", Doorkeeper revocará el token usado para actualizar cada vez.
> 
> Fue así antes de mí y presentado originalmente aquí [#769](https://github.com/doorkeeper-gem/doorkeeper/pull/769)

Hiii Argenis king
